Add FewRel Dataset

gchhablani opened the pull request:
Hi,

This PR closes this [Card](https://github.com/huggingface/datasets/projects/1#card-53285184) and Issue #1757.

I wasn't sure how to add `pid2name` along with the dataset so I added it as a separate configuration. For each (head, tail, tokens) triplet, I have created one example. I have added the dictionary key as `"relation"` in the dataset. Additionally, for `pubmed_unsupervised`, I kept `"relation":""` in the dictionary.

Please recommend better alternatives, if any.

Thanks,
Gunjan

lhoestq reviewed (commented):
Awesome thanks for adding this one !

What about adding the names directory in the default configuration ?
I don't find the pid2name config very convenient since we can't easily query the names by PID. 

Also it looks like the dummy_data.zip file is quite big (100KB). Do you think you could reduce its size a bit ? Ideally we should only keep 5 examples in the dummy file so that it's <20KB.

gchhablani commented:
Hi @lhoestq,

Sorry for the late response. What do you mean when you say "adding names to default config"? Should I handle "pid2name" in a single config?

lhoestq commented:
Yes I was thinking of having the pid2name field available in the default configuration (and therefore only have one config). What do you think ?

gchhablani commented:
Hi @lhoestq,

Sorry again, the last couple of weeks were a bit hectic for me. I am wondering how do you want me to achieve that. Using a custom BuilderConfig which takes in whether it is the regular data or pid2name?

gchhablani commented:
```
# coding=utf-8
# Copyright 2020 The HuggingFace Datasets Authors and the current dataset script contributor.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""FewRel Dataset."""

from __future__ import absolute_import, division, print_function

import json

import datasets


_CITATION = """@inproceedings{han-etal-2018-fewrel,
    title = "{F}ew{R}el: A Large-Scale Supervised Few-Shot Relation Classification Dataset with State-of-the-Art Evaluation",
    author = "Han, Xu and Zhu, Hao and Yu, Pengfei and Wang, Ziyun and Yao, Yuan and Liu, Zhiyuan and Sun, Maosong",
    booktitle = "Proceedings of the 2018 Conference on Empirical Methods in Natural Language Processing",
    month = oct # "-" # nov,
    year = "2018",
    address = "Brussels, Belgium",
    publisher = "Association for Computational Linguistics",
    url = "https://www.aclweb.org/anthology/D18-1514",
    doi = "10.18653/v1/D18-1514",
    pages = "4803--4809"
}

@inproceedings{gao-etal-2019-fewrel,
    title = "{F}ew{R}el 2.0: Towards More Challenging Few-Shot Relation Classification",
    author = "Gao, Tianyu and Han, Xu and Zhu, Hao and Liu, Zhiyuan and Li, Peng and Sun, Maosong and Zhou, Jie",
    booktitle = "Proceedings of the 2019 Conference on Empirical Methods in Natural Language Processing and the 9th International Joint Conference on Natural Language Processing (EMNLP-IJCNLP)",
    month = nov,
    year = "2019",
    address = "Hong Kong, China",
    publisher = "Association for Computational Linguistics",
    url = "https://www.aclweb.org/anthology/D19-1649",
    doi = "10.18653/v1/D19-1649",
    pages = "6251--6256"
}
"""

_DESCRIPTION = """\
FewRel is a large-scale few-shot relation extraction dataset, which contains more than one hundred relations and tens of thousands of annotated instances cross different domains.
"""

_HOMEPAGE = "https://thunlp.github.io/"

_LICENSE = "https://raw.githubusercontent.com/thunlp/FewRel/master/LICENSE"

DATA_URL = "https://raw.githubusercontent.com/thunlp/FewRel/master/data/"
_URLs = {
    "train_wiki": DATA_URL + "train_wiki.json",
    "val_nyt": DATA_URL + "val_nyt.json",
    "val_pubmed": DATA_URL + "val_pubmed.json",
    "val_semeval": DATA_URL + "val_semeval.json",
    "val_wiki": DATA_URL + "val_wiki.json",
    "pid2name": DATA_URL + "pid2name.json",
    "pubmed_unsupervised": DATA_URL + "pubmed_unsupervised.json",
}


class FewRelConfig(datasets.BuilderConfig):
    def __init__(self, *args, type=None, **kwargs):
        super().__init__(
            *args,
            name=f"{type}",
            **kwargs,
        )
        self.type = type
class FewRelDataset(datasets.GeneratorBasedBuilder):
    """The FewRelDataset."""

    VERSION = datasets.Version("1.0.0")

    BUILDER_CONFIGS = [
        FewRelConfig(type="train_wiki",version=VERSION, description="This covers the `train_wiki` part of FewRel dataset."),
        FewRelConfig(type="val_wiki", version=VERSION, description="This covers the `val_wiki` part of FewRel dataset."),
        FewRelConfig(type="val_nyt", version=VERSION, description="This covers the `val_nyt` part of FewRel dataset."),
        FewRelConfig(type="val_pubmed", version=VERSION, description="This covers the `val_pubmed` part of FewRel dataset."),
        FewRelConfig(type="val_semeval", version=VERSION, description="This covers the `val_semeval` part of FewRel dataset."),
        FewRelConfig(type="pubmed_unsupervised", version=VERSION, description="This covers the `pubmed_unsupervised` part of FewRel dataset."),

    ]

    def _info(self):
        if self.config.type in ["train_wiki","val_wiki","val_nyt"]:
            features = datasets.Features(
                {
                    "relation": datasets.Value("string"),
                    "tokens": datasets.Sequence(datasets.Value("string")),
                    "head": {
                        "text": datasets.Value("string"),
                        "type": datasets.Value("string"),
                        "indices": datasets.Sequence(datasets.Sequence(datasets.Value("int64"))),
                    },
                    "tail": {
                        "text": datasets.Value("string"),
                        "type": datasets.Value("string"),
                        "indices": datasets.Sequence(datasets.Sequence(datasets.Value("int64"))),
                    },
                    "relation_names": datasets.Sequence(datasets.Value("string"))
                    # These are the features of your dataset like images, labels ...
                }
            )
        elif self.config.type == "pubmed_unsupervised":
            features = datasets.Features(
                {
                    "tokens": datasets.Sequence(datasets.Value("string")),
                    "head": {
                        "text": datasets.Value("string"),
                        "type": datasets.Value("string"),
                        "indices": datasets.Sequence(datasets.Sequence(datasets.Value("int64"))),
                    },
                    "tail": {
                        "text": datasets.Value("string"),
                        "type": datasets.Value("string"),
                        "indices": datasets.Sequence(datasets.Sequence(datasets.Value("int64"))),
                    },
                }
            )
        else:
            features = datasets.Features(
                {
                    "relation": datasets.Value("string"),
                    "tokens": datasets.Sequence(datasets.Value("string")),
                    "head": {
                        "text": datasets.Value("string"),
                        "type": datasets.Value("string"),
                        "indices": datasets.Sequence(datasets.Sequence(datasets.Value("int64"))),
                    },
                    "tail": {
                        "text": datasets.Value("string"),
                        "type": datasets.Value("string"),
                        "indices": datasets.Sequence(datasets.Sequence(datasets.Value("int64"))),
                    },
                }
            )
        return datasets.DatasetInfo(
            # This is the description that will appear on the datasets page.
            description=_DESCRIPTION,
            # This defines the different columns of the dataset and their types
            features=features,  # Here we define them above because they are different between the two configurations
            # If there's a common (input, target) tuple from the features,
            # specify them here. They'll be used if as_supervised=True in
            # builder.as_dataset.
            supervised_keys=None,
            # Homepage of the dataset for documentation
            homepage=_HOMEPAGE,
            # License for the dataset if available
            license=_LICENSE,
            # Citation for the dataset
            citation=_CITATION,
        )

    def _split_generators(self, dl_manager):
        """Returns SplitGenerators."""
        if self.config.type in ["train_wiki", "val_wiki", "val_nyt"]:
            new_URLs = {self.config.type: _URLs[self.config.type], "pid2name": _URLs["pid2name"]}
            data_dir = dl_manager.download_and_extract(new_URLs)
            return [
                datasets.SplitGenerator(
                    name=datasets.Split(self.config.type),
                    # These kwargs will be passed to _generate_examples
                    gen_kwargs={
                        "filepath": data_dir[self.config.type],
                        "pid2name": data_dir["pid2name"]
                    },
                ),
            ]
        else:
            new_URLs = {self.config.type: _URLs[self.config.type]}
            data_dir = dl_manager.download_and_extract(new_URLs)
            return [
                datasets.SplitGenerator(
                    name=datasets.Split(self.config.type),
                    # These kwargs will be passed to _generate_examples
                    gen_kwargs={
                        "filepath": data_dir[self.config.type],
                    },
                ),
            ]

        

    def _generate_examples(self, filepath, pid2name=None):
        """ Yields examples. """
        pid2name_dict = {}
        if pid2name is not None:
            with open(pid2name, encoding="utf-8") as f:
                data = json.load(f)
            for key in list(data.keys()):
                name_1 = data[key][0]
                name_2 = data[key][1]
                pid2name_dict[key]=[name_1, name_2]
        with open(filepath, encoding="utf-8") as f:
            data = json.load(f)
            # print(data)
            if self.config.type in ["train_wiki", "val_wiki", "val_nyt"]:
                id = 0
                for key in list(data.keys()):
                    for items in data[key]:
                        tokens = items["tokens"]
                        h_0 = items["h"][0]
                        h_1 = items["h"][1]
                        h_2 = items["h"][2]
                        t_0 = items["t"][0]
                        t_1 = items["t"][1]
                        t_2 = items["t"][2]
                        id += 1
                        yield id, {
                            "relation": key,
                            "tokens": tokens,
                            "head": {"text": h_0, "type": h_1, "indices": h_2},
                            "tail": {"text": t_0, "type": t_1, "indices": t_2},
                            "relation_names": pid2name_dict[key]
                        }

            elif self.config.type =="pubmed_unsupervised":
                id = 0
                for items in data:
                    tokens = items["tokens"]
                    h_0 = items["h"][0]
                    h_1 = items["h"][1]
                    h_2 = items["h"][2]
                    t_0 = items["t"][0]
                    t_1 = items["t"][1]
                    t_2 = items["t"][2]
                    id += 1
                    yield id, {
                        "tokens": tokens,
                        "head": {"text": h_0, "type": h_1, "indices": h_2},
                        "tail": {"text": t_0, "type": t_1, "indices": t_2},
                    }            
            else:
                id = 0
                for key in list(data.keys()):
                    for items in data[key]:
                        tokens = items["tokens"]
                        h_0 = items["h"][0]
                        h_1 = items["h"][1]
                        h_2 = items["h"][2]
                        t_0 = items["t"][0]
                        t_1 = items["t"][1]
                        t_2 = items["t"][2]
                        id += 1
                        yield id, {
                            "relation": key,
                            "tokens": tokens,
                            "head": {"text": h_0, "type": h_1, "indices": h_2},
                            "tail": {"text": t_0, "type": t_1, "indices": t_2},
                        }
          
```

gchhablani commented:
Okay, I apologize, I guess I finally understand what is required.

Basically, using:

```python
few_rel = load_dataset('few_rel')
```
should give all the files. My bad. I was overthinking.
Will fix this.

lhoestq commented:
Yes that's it, sorry if that wasn't clear !

gchhablani commented:
Hi @lhoestq,

Since pid2name has different features from the rest of the files, how will I add them to the same config?

lhoestq commented:
If I understand correctly each sample in the "default" config has one relation, and each relation has corresponding names in pid2name.
Would it be possible to also include the names in the "default" configuration for each sample ? The names of one sample can be retrieved using the relation id no ?

gchhablani commented:
Yes, that can be done. But for some files, the name is already given instead of id. Only "train_wiki", "val_wiki", "val_nyc" have IDs. For others, I can set the name equal to the key.

lhoestq commented:
I think that's fine as long as we mention this processing explicitly in the dataset card.

gchhablani commented:
Hi @lhoestq,

I have added the changes. Please let me know in case of any issues.

Thanks,
Gunjan

lhoestq commented on datasets/few_rel/README.md:
```diff
@@ -0,0 +1,229 @@
+  - `indices`: a `list` containing `list` of token indices.
+- `names`: a `list` of `string` features containing relation names. For `pubmed_unsupervised` split, this is set to a `list` with an empty `string`. For `val_semeval` and `val_pubmed` split, this is set to a `list` with the `string` from the `relation` field.
+
+For `pid2name`:
+
+- `relation`: a `string` feature representing the PID of the relation.
+- `names`: a `list` of `string` features containings names for the PID.
+
```
```suggestion
```

lhoestq reviewed (approved):
Awesome thanks !

lhoestq commented on datasets/few_rel/few_rel.py:
```diff
@@ -0,0 +1,189 @@
+}
+
+
+class FewRelDataset(datasets.GeneratorBasedBuilder):
```
```suggestion
class FewRel(datasets.GeneratorBasedBuilder):
```
This must be the camel case version of the dataset script name `few_rel`

lhoestq commented on datasets/few_rel/dataset_infos.json:
```diff
@@ -0,0 +1 @@
+{"default": {"description": "FewRel is a large-scale few-shot relation extraction dataset, which contains more than one hundred relations and tens of thousands of annotated instances cross different domains.\n", "citation": "@inproceedings{han-etal-2018-fewrel,\n    title = \"{F}ew{R}el: A Large-Scale Supervised Few-Shot Relation Classification Dataset with State-of-the-Art Evaluation\",\n    aut ...
```
```suggestion
{"default": {"description": "FewRel is a large-scale few-shot relation extraction dataset, which contains more than one hundred relations and tens of thousands of annotated instances cross different domains.\n", "citation": "@inproceedings{han-etal-2018-fewrel,\n    title = \"{F}ew{R}el: A Large-Scale Supervised Few-Shot Relation Classification Dataset with State-of-the-Art Evaluation\",\n    author = \"Han, Xu and Zhu, Hao and Yu, Pengfei and Wang, Ziyun and Yao, Yuan and Liu, Zhiyuan and Sun, Maosong\",\n    booktitle = \"Proceedings of the 2018 Conference on Empirical Methods in Natural Language Processing\",\n    month = oct # \"-\" # nov,\n    year = \"2018\",\n    address = \"Brussels, Belgium\",\n    publisher = \"Association for Computational Linguistics\",\n    url = \"https://www.aclweb.org/anthology/D18-1514\",\n    doi = \"10.18653/v1/D18-1514\",\n    pages = \"4803--4809\"\n}\n\n@inproceedings{gao-etal-2019-fewrel,\n    title = \"{F}ew{R}el 2.0: Towards More Challenging Few-Shot Relation Classification\",\n    author = \"Gao, Tianyu and Han, Xu and Zhu, Hao and Liu, Zhiyuan and Li, Peng and Sun, Maosong and Zhou, Jie\",\n    booktitle = \"Proceedings of the 2019 Conference on Empirical Methods in Natural Language Processing and the 9th International Joint Conference on Natural Language Processing (EMNLP-IJCNLP)\",\n    month = nov,\n    year = \"2019\",\n    address = \"Hong Kong, China\",\n    publisher = \"Association for Computational Linguistics\",\n    url = \"https://www.aclweb.org/anthology/D19-1649\",\n    doi = \"10.18653/v1/D19-1649\",\n    pages = \"6251--6256\"\n}\n", "homepage": "https://thunlp.github.io/", "license": "https://raw.githubusercontent.com/thunlp/FewRel/master/LICENSE", "features": {"relation": {"dtype": "string", "id": null, "_type": "Value"}, "tokens": {"feature": {"dtype": "string", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}, "head": {"text": {"dtype": "string", "id": null, "_type": "Value"}, "type": {"dtype": "string", "id": null, "_type": "Value"}, "indices": {"feature": {"feature": {"dtype": "int64", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}, "length": -1, "id": null, "_type": "Sequence"}}, "tail": {"text": {"dtype": "string", "id": null, "_type": "Value"}, "type": {"dtype": "string", "id": null, "_type": "Value"}, "indices": {"feature": {"feature": {"dtype": "int64", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}, "length": -1, "id": null, "_type": "Sequence"}}, "names": {"feature": {"dtype": "string", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}}, "post_processed": null, "supervised_keys": null, "builder_name": "few_rel", "config_name": "default", "version": {"version_str": "1.0.0", "description": null, "major": 1, "minor": 0, "patch": 0}, "splits": {"train_wiki": {"name": "train_wiki", "num_bytes": 19923155, "num_examples": 44800, "dataset_name": "few_rel"}, "val_nyt": {"name": "val_nyt", "num_bytes": 1385642, "num_examples": 2500, "dataset_name": "few_rel"}, "val_pubmed": {"name": "val_pubmed", "num_bytes": 488502, "num_examples": 1000, "dataset_name": "few_rel"}, "val_semeval": {"name": "val_semeval", "num_bytes": 2646249, "num_examples": 8851, "dataset_name": "few_rel"}, "val_wiki": {"name": "val_wiki", "num_bytes": 5147348, "num_examples": 11200, "dataset_name": "few_rel"}, "pubmed_unsupervised": {"name": "pubmed_unsupervised", "num_bytes": 1117703, "num_examples": 2500, "dataset_name": "few_rel"}}, "download_checksums": {"https://raw.githubusercontent.com/thunlp/FewRel/master/data/train_wiki.json": {"num_bytes": 14298985, "checksum": "8e9e58a05d559be773cbe18a8d4a16ba17e6d397663f9f05a056f894a94b92c1"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_nyt.json": {"num_bytes": 1032006, "checksum": "b519adc61a2bab0f05178e33a3cd8f284338b635629493e9e09616ba81bb428e"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_pubmed.json": {"num_bytes": 432600, "checksum": "fc0f13b534f6664fe0d8aec231d7590313bc6510d00dbc1b6746147f6a2b46e6"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_semeval.json": {"num_bytes": 2203006, "checksum": "fd79cf35481f6596c23b92cd3ded3265481f3027a4b78d99921104f8fe632eb5"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_wiki.json": {"num_bytes": 3534147, "checksum": "76bb44c2485b6faed358a3ce040eadafcc8a8835245cd7e6afc1bef672d4a0bf"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/pid2name.json": {"num_bytes": 80278, "checksum": "eeedd25512a8e4c2bc06b78009836d27de81807bcba6ed929debd5cd713c07ea"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/pubmed_unsupervised.json": {"num_bytes": 1093301, "checksum": "fa3555941364024c2ac1b1857396515ecc65a515bc742e9da0502568fad5dd1f"}}, "download_size": 22674323, "post_processing_size": null, "dataset_size": 30708599, "size_in_bytes": 53382922}, "pid2name": {"description": "FewRel is a large-scale few-shot relation extraction dataset, which contains more than one hundred relations and tens of thousands of annotated instances cross different domains.\n", "citation": "@inproceedings{han-etal-2018-fewrel,\n    title = \"{F}ew{R}el: A Large-Scale Supervised Few-Shot Relation Classification Dataset with State-of-the-Art Evaluation\",\n    author = \"Han, Xu and Zhu, Hao and Yu, Pengfei and Wang, Ziyun and Yao, Yuan and Liu, Zhiyuan and Sun, Maosong\",\n    booktitle = \"Proceedings of the 2018 Conference on Empirical Methods in Natural Language Processing\",\n    month = oct # \"-\" # nov,\n    year = \"2018\",\n    address = \"Brussels, Belgium\",\n    publisher = \"Association for Computational Linguistics\",\n    url = \"https://www.aclweb.org/anthology/D18-1514\",\n    doi = \"10.18653/v1/D18-1514\",\n    pages = \"4803--4809\"\n}\n\n@inproceedings{gao-etal-2019-fewrel,\n    title = \"{F}ew{R}el 2.0: Towards More Challenging Few-Shot Relation Classification\",\n    author = \"Gao, Tianyu and Han, Xu and Zhu, Hao and Liu, Zhiyuan and Li, Peng and Sun, Maosong and Zhou, Jie\",\n    booktitle = \"Proceedings of the 2019 Conference on Empirical Methods in Natural Language Processing and the 9th International Joint Conference on Natural Language Processing (EMNLP-IJCNLP)\",\n    month = nov,\n    year = \"2019\",\n    address = \"Hong Kong, China\",\n    publisher = \"Association for Computational Linguistics\",\n    url = \"https://www.aclweb.org/anthology/D19-1649\",\n    doi = \"10.18653/v1/D19-1649\",\n    pages = \"6251--6256\"\n}\n", "homepage": "https://thunlp.github.io/", "license": "https://raw.githubusercontent.com/thunlp/FewRel/master/LICENSE", "features": {"relation": {"dtype": "string", "id": null, "_type": "Value"}, "names": {"feature": {"dtype": "string", "id": null, "_type": "Value"}, "length": -1, "id": null, "_type": "Sequence"}}, "post_processed": null, "supervised_keys": null, "builder_name": "few_rel", "config_name": "pid2name", "version": {"version_str": "1.0.0", "description": null, "major": 1, "minor": 0, "patch": 0}, "splits": {"pid2name": {"name": "pid2name", "num_bytes": 81607, "num_examples": 744, "dataset_name": "few_rel"}}, "download_checksums": {"https://raw.githubusercontent.com/thunlp/FewRel/master/data/train_wiki.json": {"num_bytes": 14298985, "checksum": "8e9e58a05d559be773cbe18a8d4a16ba17e6d397663f9f05a056f894a94b92c1"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_nyt.json": {"num_bytes": 1032006, "checksum": "b519adc61a2bab0f05178e33a3cd8f284338b635629493e9e09616ba81bb428e"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_pubmed.json": {"num_bytes": 432600, "checksum": "fc0f13b534f6664fe0d8aec231d7590313bc6510d00dbc1b6746147f6a2b46e6"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_semeval.json": {"num_bytes": 2203006, "checksum": "fd79cf35481f6596c23b92cd3ded3265481f3027a4b78d99921104f8fe632eb5"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/val_wiki.json": {"num_bytes": 3534147, "checksum": "76bb44c2485b6faed358a3ce040eadafcc8a8835245cd7e6afc1bef672d4a0bf"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/pid2name.json": {"num_bytes": 80278, "checksum": "eeedd25512a8e4c2bc06b78009836d27de81807bcba6ed929debd5cd713c07ea"}, "https://raw.githubusercontent.com/thunlp/FewRel/master/data/pubmed_unsupervised.json": {"num_bytes": 1093301, "checksum": "fa3555941364024c2ac1b1857396515ecc65a515bc742e9da0502568fad5dd1f"}}, "download_size": 22674323, "post_processing_size": null, "dataset_size": 81607, "size_in_bytes": 22755930}}
```
`few_rel_dataset` -> `few_rel`

lhoestq reviewed (commented):
Just doing a minor change before we merge

gchhablani commented:
Hi @lhoestq,

Thanks for fixing it and approving :)
